File: app/routers/checkout.public.js
import express from "express";

const router = express.Router();

/**
 * CONFIG
 */
const PAYPAL_MODE = process.env.PAYPAL_MODE || "sandbox";
const PAYPAL_BASE =
  PAYPAL_MODE === "live" ? "https://api-m.paypal.com" : "https://api-m.sandbox.paypal.com";

const PUBLIC_BASE_URL = process.env.PUBLIC_BASE_URL || "https://ecu-forgex-cloud.onrender.com";
const PAYPAL_CLIENT_ID = process.env.PAYPAL_CLIENT_ID;
const PAYPAL_CLIENT_SECRET = process.env.PAYPAL_CLIENT_SECRET;

/**
 * Tiny in-memory store (V1)
 * (después lo pasamos a DB real)
 */
const store = {
  orders: new Map(), // order_id -> order
};

function uid(prefix = "ORD") {
  return `${prefix}_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`.toUpperCase();
}

function priceByPatch(p) {
  const PR = {
    REV_LIMITER: 49,
    DPF_OFF: 99,
    EGR_OFF: 79,
    CAT_OFF: 79,
    ADBLUE_OFF: 89,
    TOP_SPEED_OFF: 49,
    DTC_OFF: 39,
  };
  return PR[p] ?? 0;
}

function calcTotalUSD(patches = []) {
  return patches.reduce((s, p) => s + priceByPatch(p), 0);
}

async function paypalAccessToken() {
  if (!PAYPAL_CLIENT_ID || !PAYPAL_CLIENT_SECRET) {
    throw new Error("Missing PAYPAL_CLIENT_ID / PAYPAL_CLIENT_SECRET");
  }

  const auth = Buffer.from(`${PAYPAL_CLIENT_ID}:${PAYPAL_CLIENT_SECRET}`).toString("base64");

  const r = await fetch(`${PAYPAL_BASE}/v1/oauth2/token`, {
    method: "POST",
    headers: {
      Authorization: `Basic ${auth}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: "grant_type=client_credentials",
  });

  if (!r.ok) {
    const t = await r.text().catch(() => "");
    throw new Error(`PayPal token failed (${r.status}): ${t || "no-body"}`);
  }

  const j = await r.json();
  return j.access_token;
}

async function paypalCreateOrder({ totalUSD, returnUrl, cancelUrl, description }) {
  const token = await paypalAccessToken();

  const payload = {
    intent: "CAPTURE",
    purchase_units: [
      {
        description: description || "ECU FORGE X services",
        amount: { currency_code: "USD", value: totalUSD.toFixed(2) },
      },
    ],
    application_context: {
      return_url: returnUrl,
      cancel_url: cancelUrl,
      user_action: "PAY_NOW",
      brand_name: "ECU FORGE X",
      shipping_preference: "NO_SHIPPING",
    },
  };

  const r = await fetch(`${PAYPAL_BASE}/v2/checkout/orders`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!r.ok) {
    const t = await r.text().catch(() => "");
    throw new Error(`PayPal create order failed (${r.status}): ${t || "no-body"}`);
  }

  const j = await r.json();
  const approve = (j.links || []).find((l) => l.rel === "approve")?.href;
  return { paypal_order_id: j.id, approve_url: approve };
}

/**
 * POST /public/checkout
 * body: { customer, vehicle, patches }
 * resp: { order_id, checkout_url }
 */
router.post("/public/checkout", async (req, res) => {
  try {
    const { customer, vehicle, patches } = req.body || {};

    if (!customer?.email || !customer?.name) {
      return res.status(400).json({ error: "Missing customer (name, email)" });
    }

    const patchList = Array.isArray(patches) ? patches : [];
    if (!patchList.length) {
      return res.status(400).json({ error: "No patches selected" });
    }

    const totalUSD = calcTotalUSD(patchList);
    if (totalUSD <= 0) {
      return res.status(400).json({ error: "Invalid total" });
    }

    const order_id = uid("ORD");
    const order = {
      id: order_id,
      status: "created",
      paid: false,
      download_ready: false,
      download_url: null,

      total_usd: totalUSD,
      patches: patchList,
      customer,
      vehicle,

      paypal_order_id: null,
      checkout_url: null,

      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    };

    store.orders.set(order_id, order);

    const returnUrl = `${PUBLIC_BASE_URL}/public/paypal/return?order_id=${encodeURIComponent(order_id)}`;
    const cancelUrl = `${PUBLIC_BASE_URL}/static/checkout.html?order_id=${encodeURIComponent(order_id)}&canceled=1`;

    const pp = await paypalCreateOrder({
      totalUSD,
      returnUrl,
      cancelUrl,
      description: `EFX ${patchList.join(", ")} — ${vehicle?.ecu || "ECU"}`,
    });

    order.paypal_order_id = pp.paypal_order_id;
    order.checkout_url = pp.approve_url || null;
    order.updated_at = new Date().toISOString();

    if (!order.checkout_url) {
      return res.status(500).json({ error: "PayPal approve link missing" });
    }

    return res.json({ order_id, checkout_url: order.checkout_url });
  } catch (e) {
    console.error(e);
    return res.status(500).json({ error: e?.message || "Server error" });
  }
});

/**
 * GET /public/order/:id
 */
router.get("/public/order/:id", async (req, res) => {
  const id = req.params.id;
  const o = store.orders.get(id);
  if (!o) return res.status(404).json({ error: "not_found" });

  return res.json({
    id: o.id,
    status: o.status,
    paid: o.paid,
    download_ready: o.download_ready,
    download_url: o.download_url,
    checkout_url: o.checkout_url,
    patches: o.patches,
    total_usd: o.total_usd,
    created_at: o.created_at,
  });
});

/**
 * GET /public/paypal/return
 * PayPal vuelve con ?token=<paypal_order_id>
 */
router.get("/public/paypal/return", async (req, res) => {
  const order_id = String(req.query.order_id || "");
  const token = String(req.query.token || "");
  const o = store.orders.get(order_id);

  if (!o) return res.redirect(`/static/checkout.html?order_id=${encodeURIComponent(order_id)}&err=order_not_found`);
  if (!token) return res.redirect(`/static/checkout.html?order_id=${encodeURIComponent(order_id)}&err=missing_token`);

  o.status = "approved";
  o.paypal_order_id = token;
  o.updated_at = new Date().toISOString();

  return res.redirect(`/static/checkout.html?order_id=${encodeURIComponent(order_id)}&approved=1`);
});

export default router;

File: app/routers/checkout_public.py
# app/routers/checkout_public.py
from __future__ import annotations

import os
import uuid
from datetime import datetime
from typing import Any, Dict, List

from fastapi import APIRouter, HTTPException, Request

from app.services.storage import save_order, load_order  # ✅ mismo storage que orders.py

router = APIRouter(prefix="/public", tags=["public-checkout"])

PUBLIC_BASE_URL = os.getenv("PUBLIC_BASE_URL", "https://ecu-forgex-cloud.onrender.com")

# precios base (USD)
PRICE = {
    "REV_LIMITER": 49,
    "DPF_OFF": 99,
    "EGR_OFF": 79,
    "CAT_OFF": 79,
    "ADBLUE_OFF": 89,
    "TOP_SPEED_OFF": 49,
    "DTC_OFF": 39,
}


def calc_total_usd(patches: List[str]) -> int:
    total = 0
    for p in patches:
        total += int(PRICE.get(str(p).strip(), 0))
    return total


def pick_customer_name(customer: Dict[str, Any]) -> str:
    # Wix te manda full_name + company, pero tu V1 pedía name
    name = (customer.get("full_name") or customer.get("name") or "").strip()
    if name:
        return name
    # fallback: si no hay nombre, usar company
    return (customer.get("company") or "").strip()


@router.post("/checkout")
async def public_checkout(req: Request):
    """
    Payload esperado desde Wix:
    {
      customer: {...},            # full_name, company, email, phone, address, comuna, country
      vehicle: {...},             # brand, model, serie, year, fuel, fuelDb, ecu
      selected_patches: [...],    # o patches: [...]
      currency: "USD",
      meta: {...}
    }
    """
    body = await req.json()

    customer = body.get("customer") or {}
    vehicle = body.get("vehicle") or {}
    patches = body.get("selected_patches") or body.get("patches") or []

    # -------- validar customer --------
    email = (customer.get("email") or "").strip()
    name = pick_customer_name(customer)

    if not email:
        raise HTTPException(status_code=400, detail="Missing customer.email")
    if not name:
        raise HTTPException(status_code=400, detail="Missing customer.full_name (or name)")

    # -------- validar patches --------
    if not isinstance(patches, list) or len(patches) == 0:
        raise HTTPException(status_code=400, detail="No patches selected")

    patches = [str(p).strip() for p in patches if str(p).strip()]
    if not patches:
        raise HTTPException(status_code=400, detail="No patches selected")

    total_usd = calc_total_usd(patches)
    if total_usd <= 0:
        raise HTTPException(status_code=400, detail="Invalid total (prices not found)")

    # -------- crear orden compatible con checkout.html / orders.py --------
    order_id = str(uuid.uuid4())

    order = {
        "id": order_id,
        "created_at": datetime.utcnow().isoformat(),

        # ✅ sin auth todavía: dejamos owner_email igual para trazabilidad
        "owner_email": email,

        # lo que viene desde Wix
        "customer": {
            **customer,
            "name": name,
        },
        "vehicle": vehicle,
        "selected_patches": patches,

        # precio
        "currency": "USD",
        "price_usd": float(total_usd),

        # estado inicial
        "status": "pending_payment",
        "paid": False,
        "download_ready": False,
        "download_url": None,

        # checkout (absoluto, para que Wix redirija sin dudas)
        "checkout_url": f"{PUBLIC_BASE_URL}/static/checkout.html?order_id={order_id}",

        # meta opcional
        "meta": body.get("meta") or {},
    }

    # ✅ persistir en el MISMO storage que orders.py
    save_order(order_id, order)

    return {
        "order_id": order_id,
        "checkout_url": order["checkout_url"],
        "total_usd": total_usd,
        "currency": "USD",
    }


@router.get("/order/{order_id}")
def public_order(order_id: str):
    o = load_order(order_id)
    if not o:
        raise HTTPException(status_code=404, detail="not_found")

    # ✅ respuesta pública mínima (lo que necesita analyze/checkout)
    return {
        "id": o.get("id"),
        "status": o.get("status"),
        "paid": o.get("paid"),
        "download_ready": o.get("download_ready"),
        "download_url": o.get("download_url"),
        "checkout_url": o.get("checkout_url"),
        "price_usd": o.get("price_usd"),
        "currency": o.get("currency", "USD"),
        "selected_patches": o.get("selected_patches") or o.get("patches") or [],
        "vehicle": o.get("vehicle") or {},
        "customer_email": (o.get("customer") or {}).get("email") or o.get("owner_email"),
        "created_at": o.get("created_at"),
    }


@router.post("/demo/confirm_payment/{order_id}")
def public_confirm_payment_demo(order_id: str):
    o = load_order(order_id)
    if not o:
        raise HTTPException(status_code=404, detail="not_found")

    # ✅ DEMO: pagado + descarga lista
    o["status"] = "paid"
    o["paid"] = True
    o["download_ready"] = True
    o["download_url"] = f"/download/{order_id}"

    save_order(order_id, o)

    return {"ok": True, "order_id": order_id, "download_url": o["download_url"]}
